File: pset6/readability/readability.py
from cs50 import get_string
from string import ascii_lowercase


def coleman_liau(L, S):
    # formula for coleman liau index
    # 0.0588 * L - 0.296 * S - 15.8
    return 0.0588 * L - 0.296 * S - 15.8


def print_grade(index):
    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {round(index)}")


s = get_string("Text: ")

words, sents, chars = 0, 0, 0
for c in s:
    if c in [".", "!", "?"]:
        sents += 1
    elif c == " ":
        words += 1
    elif c.lower() in ascii_lowercase:
        chars += 1

# to account for the last word        
words += 1
    
# L is the average number of letters per 100 words in the text, and S is the average number of sentences per 100 words in the text.
L = chars / words * 100
S = sents / words * 100

index = coleman_liau(L, S)
print_grade(index)

File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

// function declaration to be defined later
float coleman_liau(float l, float s);

int main(void)
{
    // reading input string
    string s = get_string("Text: ");
    
    // initializations, words is initialized to 1 as the last word is not counted
    int words = 1;
    int characters = 0;
    int sentences = 0;
    
    // for loop to count words, characters and sentences
    for (int i = 0, length = strlen(s); i < length; i++)
    {
        if (s[i] == ' ') 
        {
            words++;
        } 
        else if (s[i] == '.' || s[i] == '?' || s[i] == '!')
        {
            sentences++;
        } 
        else if ((s[i] >= 'a' && s[i] <= 'z') || (s[i] >= 'A' && s[i] <= 'Z'))
        {
            characters++;
        }
        else 
        {
            continue;   
        }
    }
    
    // calculate readability index
    float L = (float) characters / words * 100;
    float S = (float) sentences / words * 100;
    float index = coleman_liau(L, S);
    
    // printing results
    if (index < 1) 
    {
        printf("Before Grade 1");
    }
    else if (index > 16)
    {
        printf("Grade 16+");
    }
    else 
    {
        printf("Grade %i", (int) round(index));
    }
    printf("\n");
}

float coleman_liau(float L, float S) 
{
    return 0.0588 * L - 0.296 * S - 15.8;
}